### PR TITLE
getting_started in Golang

### PR DESCRIPTION
**Pull request by patilsnr**
adds getting_started sample flow in golang using mqtt5

### CODE REVIEW

File: scenarios/getting_started/go/getting_started.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+	defer stop()
+
+	cfg := &tls.Config{
+		Certificates: []tls.Certificate{cert},
+	}
+
+	clientConfig := autopaho.ClientConfig{
+		BrokerUrls: []*url.URL{u},
+		TlsCfg:     cfg,
+		KeepAlive:  30,
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:48]: KeepAlive should be configurable

**Review by rido-min — CHANGES_REQUESTED** [review]
- This sample should be aligned with other langauges, and refactor common code to mqttclients folder to enable reuse for future samples
- Consistency with default values defined 
- Do not use autopaho
- Update READMEs
- Include instructions for mosquitto (including how to configure the CA)

File: scenarios/getting_started/go/getting_started.go [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+	KeepAlive uint16
+	ClientId  string
+	Username  string
+	KeyFile   string
+}
+
+func main() {
+	// Load connection settings
+	godotenv.Load("../.env")
+	cs := MqttConnectionSettings{
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:30]: we will repeat the same code in all samples. Can you move this to /mqttclients ?

File: scenarios/getting_started/go/getting_started.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+)
+
+type MqttConnectionSettings struct {
+	Hostname  string
+	TcpPort   int
+	UseTls    bool
+	CaFile    string
+	KeepAlive uint16
+	ClientId  string
+	Username  string
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:23]: why username but no password?

File: scenarios/getting_started/go/getting_started.go [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+	// Load connection settings
+	godotenv.Load("../.env")
+	cs := MqttConnectionSettings{
+		Hostname:  os.Getenv("MQTT_HOST_NAME"),
+		TcpPort:   8883,
+		Username:  os.Getenv("MQTT_USERNAME"),
+		ClientId:  os.Getenv("MQTT_CLIENT_ID"),
+		CaFile:    os.Getenv("MQTT_CERT_FILE"),
+		KeepAlive: 30,
+		KeyFile:   os.Getenv("MQTT_KEY_FILE"),
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:37]: I miss some initialization, such as UseTls, or tcpPort

File: scenarios/getting_started/go/getting_started.go [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	// Load certificates
+	fmt.Println("Loading certificates")
+	cert, err := tls.LoadX509KeyPair(fmt.Sprintf("../%s", cs.CaFile), fmt.Sprintf("../%s", cs.KeyFile))
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:42]: mmhhh LoadX509KeyPair seems needed to load the private key, not needed here. Is there any LoadPem? or similar

File: scenarios/getting_started/go/getting_started.go [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+		TcpPort:   8883,
+		Username:  os.Getenv("MQTT_USERNAME"),
+		ClientId:  os.Getenv("MQTT_CLIENT_ID"),
+		CaFile:    os.Getenv("MQTT_CERT_FILE"),
+		KeepAlive: 30,
+		KeyFile:   os.Getenv("MQTT_KEY_FILE"),
+	}
+
+	// Load certificates
+	fmt.Println("Loading certificates")
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:41]: we should load the CAFile only if the envvar was set

File: scenarios/getting_started/go/getting_started.go [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+	cert, err := tls.LoadX509KeyPair(fmt.Sprintf("../%s", cs.CaFile), fmt.Sprintf("../%s", cs.KeyFile))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	cfg := &tls.Config{
+		Certificates: []tls.Certificate{cert},
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:48]: we need to load client certificate if specified

File: scenarios/getting_started/go/getting_started.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+	}
+
+	fmt.Println("Dialing Eventgrid")
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:51]: Dialing?

File: scenarios/getting_started/go/getting_started.go [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+	conn, err := tls.Dial("tcp", fmt.Sprintf("%s:%d", cs.Hostname, cs.TcpPort), cfg)
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:52]: what if UseTls was set to false?

File: scenarios/getting_started/go/getting_started.go [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+			} else {
+				fmt.Printf("server requested disconnect; reason code: %d\n", d.ReasonCode)
+			}
+		},
+	})
+
+	cp := &paho.Connect{
+		KeepAlive:    cs.KeepAlive,
+		ClientID:     cs.ClientId,
+		CleanStart:   true,
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:79]: why not configuring CleanStart ?

File: scenarios/getting_started/go/getting_started.go [hunk 11]
```diff
@@ -0,0 +1,110 @@
...
+
+func main() {
+	// Load connection settings
+	godotenv.Load("../.env")
+	cs := MqttConnectionSettings{
+		Hostname:  os.Getenv("MQTT_HOST_NAME"),
+		TcpPort:   8883,
+		Username:  os.Getenv("MQTT_USERNAME"),
+		ClientId:  os.Getenv("MQTT_CLIENT_ID"),
+		CaFile:    os.Getenv("MQTT_CERT_FILE"),
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:35]: do not confuse CaFile with CertFile

**Review by rido-min — CHANGES_REQUESTED** [review]
missing some envvars
move code to mosquitto clients to enable future reuse
Clarify when to use client certs vs. CaFile

Have you tested with Mosquitto ?

We also need to update all the READMEs

File: scenarios/getting_started/go/getting_started.go [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+	KeepAlive uint16
+	ClientId  string
+	Username  string
+	KeyFile   string
+}
+
+func main() {
+	// Load connection settings
+
+	cs := ConnectionSettings.loadConnectionSettings("../.env")
```
Comment (rido-min) [inline scenarios/getting_started/go/getting_started.go:29]: we should not hardcode the path to the .env file, or we accept as an argument, or we try to load from current dir.

**Review by rido-min — CHANGES_REQUESTED** [review]
This is looking great !!, 

although I think it still missing tests, and README updates (not only the Getting Started README).

Have you tested in mosquitto and EG?
Is the codespaces devContainer ready for GO?

**Review by rido-min — CHANGES_REQUESTED** [review]
any chance to unit test connectionsettings.go?
I'm missing some logic to enable/disable TLS

Have you tried with mosquitto?

**Comment by patilsnr** [comment]
> any chance to unit test connectionsettings.go? I'm missing some logic to enable/disable TLS
> 
> Have you tried with mosquitto?

I'm still addressing your comments and pushing periodically. I will let you know when this is ready for a proper review.

**Comment by rido-min** [comment]
> > any chance to unit test connectionsettings.go? I'm missing some logic to enable/disable TLS
> > Have you tried with mosquitto?
> 
> I'm still addressing your comments and pushing periodically. I will let you know when this is ready for a proper review.

you can set the PR as draft until it's ready

File: .gitignore [hunk 13]
```diff
@@ -8,6 +8,7 @@
 *.pem
 *.crt
 *.env
+*.exe
```
Comment (rido-min) [inline .gitignore:11]: this will only work on Windows, not linux (no extension)

We should find a way to exclude the output folder.

**Review by rido-min — APPROVED** [review]
LGTM

**Pull request merged**